packages/protocol/contracts/bridge/libs/LibBridgeRelease.sol

```diff
@@ -56,37 +66,73 @@ library LibBridgeRelease {
 
         bytes32 msgHash = message.hashMessage();
 
-        if (state.etherReleased[msgHash] == true) {
-            revert B_ETHER_RELEASED_ALREADY();
-        }
-
         if (
             !LibBridgeStatus.isMessageFailed(
                 resolver, msgHash, message.destChainId, proof
             )
         ) {
             revert B_MSG_NOT_FAILED();
         }
+        
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.FULLY_RECALLED
+        ){
+            // Both ether and tokens are released
+            revert B_TOKENS_RELEASED_ALREADY();
+        }
 
-        state.etherReleased[msgHash] = true;
+        uint256 releaseAmount;
 
-        uint256 releaseAmount = message.depositValue + message.callValue;
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.NOT_RECALLED
+        ) {
+            // Release ETH first
+            state.recallStatus[msgHash] = LibBridgeData.RecallStatus.ETH_RELEASED;
 
-        if (releaseAmount > 0) {
-            address ethVault = resolver.resolve("ether_vault", true);
-            // if on Taiko
-            if (ethVault != address(0)) {
-                EtherVault(payable(ethVault)).releaseEther(
-                    message.owner, releaseAmount
-                );
-            } else {
-                // if on Ethereum
-                (bool success,) = message.owner.call{ value: releaseAmount }("");
-                if (!success) {
-                    revert B_FAILED_TRANSFER();
+            releaseAmount = message.depositValue + message.callValue;
+
+            if (releaseAmount > 0) {
+                address ethVault = resolver.resolve("ether_vault", true);
+                // if on Taiko
+                if (ethVault != address(0)) {
+                    EtherVault(payable(ethVault)).releaseEther(
+                        message.owner, releaseAmount
+                    );
+                } else {
+                    // if on Ethereum
+                    (bool success,) = message.owner.call{ value: releaseAmount }("");
+                    if (!success) {
+                        revert B_FAILED_TRANSFER();
+                    }
                 }
             }
         }
+        //2nd stage is releasing the tokens
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.ETH_RELEASED 
+        ) {
+            if( message.to == address(0)
+                || !_isContract(message.sender)
+                || !IERC165Upgradeable(message.sender).supportsInterface(VAULT_INTERFACE_ID) 
```

Comment:
You cannot call `IERC165Upgradeable(message.sender)` here as this will revert if the message.sender does not have `supportsInterface` function, instead, you should define a private function `isRecallableMessageSender` and inside the function, use try-catch for `supportsInterface`...

one example:

```solidity
if (to.isContract()) {
            try IERC1155Receiver(to).onERC1155Received(operator, from, id, amount, data) returns (bytes4 response) {
                if (response != IERC1155Receiver.onERC1155Received.selector) {
                    revert("ERC1155: ERC1155Receiver rejected tokens");
                }
            } catch Error(string memory reason) {
                revert(reason);
            } catch {
                revert("ERC1155: transfer to non-ERC1155Receiver implementer");
            }
        }
```
Also require the onMessageRecalled to return the selector!